### ISSUE TITLE
Paths are not stitched together correctly

### ISSUE THREAD

**Issue by kleinfreund** [11ty/eleventy#231]
The biggest source of frustration with Eleventy is the in my opinion overly restrictive directory structure. Especially the global includes and data directories being relative to the input directory is a hinderance and I would like that to go away in the future. That’s another discussion though.

Hacking around this, I found that setting `dir.includes` to the empty string (i.e. `""`) allows one to include from across the whole input directory instead of just from the includes directory. Setting the includes directory like this effectively makes the input directory the includes directory. **Almost**.

When setting Eleventy’s directories via the configuration, the resulting paths are just concatenated. This is no problem most of the time. In my case, Eleventy considers `dir.input` the input directory and `dir.input/` the includes directory which most of the time will resolve to the same directory.

I just found an exception: Pagination. Instead of creating a page ULR like `/1/`, I currently end up with `//1/` which browsers treat as an absolute URL with an unspecified scheme. It’s actually pretty interesting: `//1/` is resolves to the URL `0.0.0.1` because the part after `//` is a number and used as the last part for the address `0.0.0.1`.

**TL;DR**: Eleventy directories aren’t always stitched together correctly. They should be sanitized before storing them in the computed configuration. I compiled a list of combinations of input and includes directories that *could* cause some problems.

| dir.input | dir.includes | Computed includes dir | Expected includes dir |
|---------|------------|----------------|--------------------|
| `'src'` | `'_includes'`       | `'src/includes'`       | `'src/includes'`            |
| `'src'` | `''`       | `'src/'`       | `'src'`            |
| `'src'` | `'/'`      | `'src//'`      | `'src'`            |
| `'src'` | `'.'`      | `'src/.'`      | `'src'`            |
| `'src'` | `'./'`      | `'src/./'`      | `'src'`            |
| `'src'` | `'../'`    | `'src/../'`    | Should error       |
| `'src'` | `'../../'` | `'src/../../'` | Should super error |

Using `../` should be flat-out forbidden to avoid accessing files outside of Eleventy’s realm. If people really want this, they should use symbolic links.

Generally, I would avoid storing anything with a trailing slash. This way, it can always be appended if necessary and no double-slash scenarios are created accidantelly.

When concatenating paths, double-slash segments should be *smoosh*ed to only one slash. Also, segments like `/./` should be `/`.

I’m sure there is an <del>app</del><ins>package</ins> for that. Also, I’m back from vacation. :sunglasses:

**Comment by kleinfreund** [comment]
I’d like to work on this. I found the place where the paths are concatenated in the [`EleventyFiles.js`](https://github.com/11ty/eleventy/blob/master/src/EleventyFiles.js#L21) file. I propose a change like this using [Node.js’ `path.join`](https://nodejs.org/api/path.html#path_path_join_paths):

```diff
- this.includesDir = this.inputDir + "/" + this.config.dir.includes;
+ this.includesDir = path.join(this.inputDir, this.config.dir.includes).replace(/\/+$/, "");
```

This also removes all trailing slashes from the path.

Additionally, I’d like to check for `..` segments, but this might be more correct to do earlier, preferably as soon as Eleventy sees the includes directory. I think, `..` segments should produce an error immediately, but maybe are warning is enough. What do you think about this?

As for the global data directory, the correct place for this change would probably be in [`TemplateData.js`](https://github.com/11ty/eleventy/blob/master/src/TemplateData.js#L32-L35).

Doing a quick search across the project reveals that there are a lot of instances where paths are concatenated with the pattern `parentDir + "/" + childDir`. It could be worthwhile to review these as well.

**Comment by zachleat** [comment]
I need a bit more time to review this (it looks great on first glance) but I did want to leave a quick comment to make sure you’ve seen `TemplatePath.normalize` function https://github.com/11ty/eleventy/blob/master/src/TemplatePath.js#L62 I think there is some overlap there in what you’re suggesting.

If not, I’d love a method that works similarly, takes a bunch of args and removes duplicate slashes, etc. Makes for easy testing too.

One other thing that comes to mind is the Template functions used to output the `url`, which should definitely keep trailing slashes per discussion in #213.

**Comment by kleinfreund** [comment]
I haven’t seen that normalizing function. I thought about having a utility normalizing function for Eleventy paths that would most likely call `path.join` like `TemplatePath.normalize` does.

Note that the comment there is not correct. `path.join` doesn’t remove trailing slashes.

Also, very valid point with regards to permalinks: If removing a trailing slash there should change Eleventy’s general behavior, we should document it and take care of it.

We can also make this two separate changes: First, normalizing includes and data directories. Second, evaluating all other instances of path concatenation and fixing potential problem cases.

**Comment by kleinfreund** [comment]
#186 is somewhat related. Both the input and the output directory should be located inside Eleventy’s project/command directory. Also, they should not be descendants of each other.

**Comment by rendall** [comment]
The nodejs path library is pretty nifty.  In particular, the 'normalize' function would smooth all of those paths out without extra dependencies.

https://nodejs.org/api/path.html#path_path_normalize_path

**Comment by kleinfreund** [comment]
@rendall The proposed change above uses `path.join` which normalizes the paths already.

**Comment by kleinfreund** [comment]
@zachleat What do you think now? I think we should implement a utility function `normalizedPath` that takes an arbitrary amount of path segments, joins them together and normalizes them. This might be just calling `path.join`, but allows to transition to a different mechanism later on. Every Eleventy directory/path should only ever be stored after calling this function.

**Comment by zachleat** [comment]
Yes I definitely think we need some cleanup here @kleinfreund. Note Eleventy’s `TemplatePath` class has a lot of this already built in, it just needs to be applied consistently. Check out the tests to see how it works https://github.com/11ty/eleventy/blob/master/test/TemplatePathTest.js#L1
